packages/sdk/src/protocols/peer_manager.ts

```diff
@@ -0,0 +1,113 @@
+import { Peer, PeerId } from "@libp2p/interface";
+import { ConnectionManager, getHealthManager } from "@waku/core";
+import { BaseProtocol } from "@waku/core/lib/base_protocol";
+import { IHealthManager } from "@waku/interfaces";
+import { Logger } from "@waku/utils";
+import { Mutex } from "async-mutex";
```

Comment:
have you considered using `connection.tags` to lock connection for usage? 
I think this way we can ease implementation and avoid usage of `async-mutext` 

this would mean that:
- peers that are used can always be retrieved by `connections.filter(c => c.tag["USED"])` and no need to support any copy of it on our side;
- subscriptions/peer manager can be reactive to change by using `peer:connect` event;

Reply:
I agree with the idea and align with it -- I'm unsure of the performance or scalability/extensibility pitfalls it may take. Do you have a perspective on this?

Reply:
To me it seems to be rather safe choice to make from performance PoV:
- find / retrieval of connections will be just an iteration over amount of connections (`O(n)`) which is not growing that much because we have limited amount of them;
- there is no `async` operations involved for iteration. apart from, maybe, reading from `PeerStore` at some point - this can be done either `lazy` or once when we absolutely need `Peer` object;
- and at the end - I wouldn't think about performance before we actually hit problems with that.

Most important part is to have coherence and consistency with underlying `libp2p`.

Reply:
will be an important follow up https://github.com/waku-org/js-waku/issues/2167